x-pack/plugins/security_solution/server/lib/detection_engine/rule_execution_log/rule_execution_logger/logger.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { sum } from 'lodash';
+import { Logger } from 'kibana/server';
+
+import {
+  RuleExecutionStatus,
+  RuleExecutionStatusOrder,
+  RuleExecutionMetrics,
+} from '../../../../../common/detection_engine/schemas/common';
+
+import { RuleAlertType } from '../../rules/types';
+import { ExtMeta } from '../types';
+import { truncateMessage } from '../utils/normalization';
+
+import { IRuleExecutionEventsWriter } from '../rule_execution_events/events_writer';
+import { IRuleExecutionInfoSavedObjectsClient } from '../rule_execution_info/saved_objects_client';
+import { IRuleExecutionLogger, StatusChangeArgs } from './logger_interface';
+
+export const createRuleExecutionLogger = (
+  rule: RuleAlertType,
+  spaceId: string,
+  executionEventsWriter: IRuleExecutionEventsWriter,
+  executionInfoClient: IRuleExecutionInfoSavedObjectsClient,
+  logger: Logger
+): IRuleExecutionLogger => {
+  const ruleId = rule.id;
+  const ruleName = rule.name;
+  const ruleType = rule.alertTypeId;
+
+  const ruleExecutionLogger: IRuleExecutionLogger = {
+    async logStatusChange(args) {
+      try {
+        const normalizedArgs = normalizeStatusChangeArgs(args);
+        await Promise.all([
+          writeStatusChangeToSavedObjects(normalizedArgs),
+          writeStatusChangeToEventLog(normalizedArgs),
+        ]);
+      } catch (e) {
+        const logMessage = 'Error logging rule execution status change';
+        const logAttributes = `status: "${args.newStatus}", rule id: "${ruleId}", rule name: "${ruleName}"`;
+        const logReason = e instanceof Error ? `${e.stack}` : `${e}`;
```

Comment:
`e.stack` includes both the message and the stack, while `e.message` contains only the message.

Reply:
Now I have more thoughts behind this 🙂 

`IRuleExecutionLogger.logStatusChange` catches an exception and doesn't re-throw it, this is why it's important to log its stack trace along with the original message and correlation ids. Without the stack it would be harder to find the exact place where the exception originated from. The implementation of `logReason` is a little bit adjusted:

```ts
    async logStatusChange(args) {
      try {
        const normalizedArgs = normalizeStatusChangeArgs(args);
        await Promise.all([
          writeStatusChangeToSavedObjects(normalizedArgs),
          writeStatusChangeToEventLog(normalizedArgs),
        ]);
      } catch (e) {
        const logMessage = 'Error logging rule execution status change';
        const logAttributes = `status: "${args.newStatus}", rule id: "${ruleId}", rule name: "${ruleName}"`;
        const logReason = e instanceof Error ? e.stack ?? e.message : String(e);
        const logMeta: ExtMeta = {
          rule: {
            id: ruleId,
            name: ruleName,
            type: ruleType,
            execution: {
              status: args.newStatus,
            },
          },
          kibana: {
            spaceId,
          },
        };

        logger.error<ExtMeta>(`${logMessage}; ${logAttributes}; ${logReason}`, logMeta);
      }
    },
  };
```

Methods of `IRuleExecutionInfoSavedObjectsClient` re-throw exceptions, so I did not include the stack trace (only the message) to save some space in the logs:

```ts
    async getOneByRuleId(ruleId) {
      try {
        const id = getRuleExecutionInfoId(ruleId);
        return await soClient.get<RuleExecutionInfoAttributes>(RULE_EXECUTION_INFO_TYPE, id);
      } catch (e) {
        if (SavedObjectsErrorHelpers.isNotFoundError(e)) {
          return null;
        }

        const logMessage = 'Error fetching rule execution info saved object';
        const logAttributes = `rule id: "${ruleId}"`;
        const logReason = e instanceof Error ? e.message : String(e);
        const logMeta: ExtMeta = {
          rule: { id: ruleId },
        };

        logger.error<ExtMeta>(`${logMessage}; ${logAttributes}; ${logReason}`, logMeta);
        throw e;
      }
    },
```